{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7c55c698",
   "metadata": {},
   "source": [
    "# Codon Usage"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c9de65e",
   "metadata": {},
   "source": [
    "### Abstract: \n",
    "DNA codon usage frequencies of a large sample of diverse biological organisms from different taxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0e3c3e1",
   "metadata": {},
   "source": [
    "### Source:\n",
    "Bohdan Khomtchouk, Ph.D. University of Chicago, Department of Medicine, Section of Computational Biomedicine and Biomedical Data Science.\n",
    "\n",
    "#### Date Donated: 2020-10-03"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3aeac770",
   "metadata": {},
   "source": [
    "### Attribute Information:\n",
    "\n",
    "- Column 1: Kingdom \n",
    "- Column 2: DNAtype \n",
    "- Column 3: SpeciesID \n",
    "- Column 4: Ncodons \n",
    "- Column 5: SpeciesName \n",
    "- Columns 6-69: codon (header: nucleotide bases; entries: frequency of usage (5 digit floating point number)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7311cb1",
   "metadata": {},
   "source": [
    "### Nucleotides bases\n",
    "\n",
    "- A : Adenine\n",
    "- C : Cytosine\n",
    "- G : Guanine\n",
    "- U : Uracil"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a20d4a",
   "metadata": {},
   "source": [
    "## Problem Statement :\n",
    "\n",
    "### Classifying different species into their Kingdom types on the basis of codon frequency usage as input features with Genomic DNA type.\n",
    "\n",
    "#### Class labels: \n",
    "1. Virus \n",
    "2. Bacteria \n",
    "3. Plants\n",
    "4. Vertebrates\n",
    "5. Invertebrates"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e20eff5",
   "metadata": {},
   "source": [
    "## Importing libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2280fc7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(action='ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ced2fe1f",
   "metadata": {},
   "source": [
    "## Reading dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bba060a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_dir = Path('../data')\n",
    "data_path = data_dir / 'codon_usage_dataset_processed.csv'\n",
    "\n",
    "df = pd.read_csv(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b7124b76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "      <th>SpeciesID</th>\n",
       "      <th>Ncodons</th>\n",
       "      <th>SpeciesName</th>\n",
       "      <th>UUU</th>\n",
       "      <th>UUC</th>\n",
       "      <th>UUA</th>\n",
       "      <th>UUG</th>\n",
       "      <th>CUU</th>\n",
       "      <th>CUC</th>\n",
       "      <th>...</th>\n",
       "      <th>CGG</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGG</th>\n",
       "      <th>GAU</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>UAA</th>\n",
       "      <th>UAG</th>\n",
       "      <th>UGA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vrl</td>\n",
       "      <td>100217</td>\n",
       "      <td>1995</td>\n",
       "      <td>Epizootic haematopoietic necrosis virus</td>\n",
       "      <td>0.01654</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>0.00351</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.03208</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00451</td>\n",
       "      <td>0.01303</td>\n",
       "      <td>0.03559</td>\n",
       "      <td>0.01003</td>\n",
       "      <td>0.04612</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.04361</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vrl</td>\n",
       "      <td>100220</td>\n",
       "      <td>1474</td>\n",
       "      <td>Bohle iridovirus</td>\n",
       "      <td>0.02714</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.02849</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.01696</td>\n",
       "      <td>0.03596</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.04545</td>\n",
       "      <td>0.01560</td>\n",
       "      <td>0.04410</td>\n",
       "      <td>0.00271</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>vrl</td>\n",
       "      <td>100755</td>\n",
       "      <td>4862</td>\n",
       "      <td>Sweet potato leaf curl virus</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.02180</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.01543</td>\n",
       "      <td>0.00782</td>\n",
       "      <td>0.01111</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00596</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.02489</td>\n",
       "      <td>0.03126</td>\n",
       "      <td>0.02036</td>\n",
       "      <td>0.02242</td>\n",
       "      <td>0.02468</td>\n",
       "      <td>0.00391</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>vrl</td>\n",
       "      <td>100880</td>\n",
       "      <td>1915</td>\n",
       "      <td>Northern cereal mosaic virus</td>\n",
       "      <td>0.01775</td>\n",
       "      <td>0.02245</td>\n",
       "      <td>0.01619</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.01567</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00366</td>\n",
       "      <td>0.01410</td>\n",
       "      <td>0.01671</td>\n",
       "      <td>0.03760</td>\n",
       "      <td>0.01932</td>\n",
       "      <td>0.03029</td>\n",
       "      <td>0.03446</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.00157</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vrl</td>\n",
       "      <td>100887</td>\n",
       "      <td>22831</td>\n",
       "      <td>Soil-borne cereal mosaic virus</td>\n",
       "      <td>0.02816</td>\n",
       "      <td>0.01371</td>\n",
       "      <td>0.00767</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0.01380</td>\n",
       "      <td>0.00548</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00604</td>\n",
       "      <td>0.01494</td>\n",
       "      <td>0.01734</td>\n",
       "      <td>0.04148</td>\n",
       "      <td>0.02483</td>\n",
       "      <td>0.03359</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00044</td>\n",
       "      <td>0.00131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 68 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Kingdom  SpeciesID  Ncodons                              SpeciesName  \\\n",
       "0     vrl     100217     1995  Epizootic haematopoietic necrosis virus   \n",
       "1     vrl     100220     1474                         Bohle iridovirus   \n",
       "2     vrl     100755     4862             Sweet potato leaf curl virus   \n",
       "3     vrl     100880     1915             Northern cereal mosaic virus   \n",
       "4     vrl     100887    22831           Soil-borne cereal mosaic virus   \n",
       "\n",
       "       UUU      UUC      UUA      UUG      CUU      CUC  ...      CGG  \\\n",
       "0  0.01654  0.01203  0.00050  0.00351  0.01203  0.03208  ...  0.00451   \n",
       "1  0.02714  0.01357  0.00068  0.00678  0.00407  0.02849  ...  0.00136   \n",
       "2  0.01974  0.02180  0.01357  0.01543  0.00782  0.01111  ...  0.00596   \n",
       "3  0.01775  0.02245  0.01619  0.00992  0.01567  0.01358  ...  0.00366   \n",
       "4  0.02816  0.01371  0.00767  0.03679  0.01380  0.00548  ...  0.00604   \n",
       "\n",
       "       AGA      AGG      GAU      GAC      GAA      GAG      UAA      UAG  \\\n",
       "0  0.01303  0.03559  0.01003  0.04612  0.01203  0.04361  0.00251  0.00050   \n",
       "1  0.01696  0.03596  0.01221  0.04545  0.01560  0.04410  0.00271  0.00068   \n",
       "2  0.01974  0.02489  0.03126  0.02036  0.02242  0.02468  0.00391  0.00000   \n",
       "3  0.01410  0.01671  0.03760  0.01932  0.03029  0.03446  0.00261  0.00157   \n",
       "4  0.01494  0.01734  0.04148  0.02483  0.03359  0.03679  0.00000  0.00044   \n",
       "\n",
       "       UGA  \n",
       "0  0.00000  \n",
       "1  0.00000  \n",
       "2  0.00144  \n",
       "3  0.00000  \n",
       "4  0.00131  \n",
       "\n",
       "[5 rows x 68 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d9b5bd77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8657, 68)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "eb7dc56c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 8657 entries, 0 to 8656\n",
      "Data columns (total 68 columns):\n",
      " #   Column       Non-Null Count  Dtype  \n",
      "---  ------       --------------  -----  \n",
      " 0   Kingdom      8657 non-null   object \n",
      " 1   SpeciesID    8657 non-null   int64  \n",
      " 2   Ncodons      8657 non-null   int64  \n",
      " 3   SpeciesName  8657 non-null   object \n",
      " 4   UUU          8657 non-null   float64\n",
      " 5   UUC          8657 non-null   float64\n",
      " 6   UUA          8657 non-null   float64\n",
      " 7   UUG          8657 non-null   float64\n",
      " 8   CUU          8657 non-null   float64\n",
      " 9   CUC          8657 non-null   float64\n",
      " 10  CUA          8657 non-null   float64\n",
      " 11  CUG          8657 non-null   float64\n",
      " 12  AUU          8657 non-null   float64\n",
      " 13  AUC          8657 non-null   float64\n",
      " 14  AUA          8657 non-null   float64\n",
      " 15  AUG          8657 non-null   float64\n",
      " 16  GUU          8657 non-null   float64\n",
      " 17  GUC          8657 non-null   float64\n",
      " 18  GUA          8657 non-null   float64\n",
      " 19  GUG          8657 non-null   float64\n",
      " 20  GCU          8657 non-null   float64\n",
      " 21  GCC          8657 non-null   float64\n",
      " 22  GCA          8657 non-null   float64\n",
      " 23  GCG          8657 non-null   float64\n",
      " 24  CCU          8657 non-null   float64\n",
      " 25  CCC          8657 non-null   float64\n",
      " 26  CCA          8657 non-null   float64\n",
      " 27  CCG          8657 non-null   float64\n",
      " 28  UGG          8657 non-null   float64\n",
      " 29  GGU          8657 non-null   float64\n",
      " 30  GGC          8657 non-null   float64\n",
      " 31  GGA          8657 non-null   float64\n",
      " 32  GGG          8657 non-null   float64\n",
      " 33  UCU          8657 non-null   float64\n",
      " 34  UCC          8657 non-null   float64\n",
      " 35  UCA          8657 non-null   float64\n",
      " 36  UCG          8657 non-null   float64\n",
      " 37  AGU          8657 non-null   float64\n",
      " 38  AGC          8657 non-null   float64\n",
      " 39  ACU          8657 non-null   float64\n",
      " 40  ACC          8657 non-null   float64\n",
      " 41  ACA          8657 non-null   float64\n",
      " 42  ACG          8657 non-null   float64\n",
      " 43  UAU          8657 non-null   float64\n",
      " 44  UAC          8657 non-null   float64\n",
      " 45  CAA          8657 non-null   float64\n",
      " 46  CAG          8657 non-null   float64\n",
      " 47  AAU          8657 non-null   float64\n",
      " 48  AAC          8657 non-null   float64\n",
      " 49  UGU          8657 non-null   float64\n",
      " 50  UGC          8657 non-null   float64\n",
      " 51  CAU          8657 non-null   float64\n",
      " 52  CAC          8657 non-null   float64\n",
      " 53  AAA          8657 non-null   float64\n",
      " 54  AAG          8657 non-null   float64\n",
      " 55  CGU          8657 non-null   float64\n",
      " 56  CGC          8657 non-null   float64\n",
      " 57  CGA          8657 non-null   float64\n",
      " 58  CGG          8657 non-null   float64\n",
      " 59  AGA          8657 non-null   float64\n",
      " 60  AGG          8657 non-null   float64\n",
      " 61  GAU          8657 non-null   float64\n",
      " 62  GAC          8657 non-null   float64\n",
      " 63  GAA          8657 non-null   float64\n",
      " 64  GAG          8657 non-null   float64\n",
      " 65  UAA          8657 non-null   float64\n",
      " 66  UAG          8657 non-null   float64\n",
      " 67  UGA          8657 non-null   float64\n",
      "dtypes: float64(64), int64(2), object(2)\n",
      "memory usage: 4.5+ MB\n"
     ]
    }
   ],
   "source": [
    "df.info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de374eda",
   "metadata": {},
   "source": [
    "### Dividing dataset into features and labels dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3593ad79",
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = df[['Kingdom']].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0bd856d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Kingdom</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>vrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>vrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>vrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>vrl</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>vrl</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Kingdom\n",
       "0     vrl\n",
       "1     vrl\n",
       "2     vrl\n",
       "3     vrl\n",
       "4     vrl"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "534bf13e",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = df.drop(columns=['Kingdom','SpeciesID','SpeciesName'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "79ef6bbf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Ncodons</th>\n",
       "      <th>UUU</th>\n",
       "      <th>UUC</th>\n",
       "      <th>UUA</th>\n",
       "      <th>UUG</th>\n",
       "      <th>CUU</th>\n",
       "      <th>CUC</th>\n",
       "      <th>CUA</th>\n",
       "      <th>CUG</th>\n",
       "      <th>AUU</th>\n",
       "      <th>...</th>\n",
       "      <th>CGG</th>\n",
       "      <th>AGA</th>\n",
       "      <th>AGG</th>\n",
       "      <th>GAU</th>\n",
       "      <th>GAC</th>\n",
       "      <th>GAA</th>\n",
       "      <th>GAG</th>\n",
       "      <th>UAA</th>\n",
       "      <th>UAG</th>\n",
       "      <th>UGA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1995</td>\n",
       "      <td>0.01654</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>0.00351</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.03208</td>\n",
       "      <td>0.00100</td>\n",
       "      <td>0.04010</td>\n",
       "      <td>0.00551</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00451</td>\n",
       "      <td>0.01303</td>\n",
       "      <td>0.03559</td>\n",
       "      <td>0.01003</td>\n",
       "      <td>0.04612</td>\n",
       "      <td>0.01203</td>\n",
       "      <td>0.04361</td>\n",
       "      <td>0.00251</td>\n",
       "      <td>0.00050</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1474</td>\n",
       "      <td>0.02714</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.00678</td>\n",
       "      <td>0.00407</td>\n",
       "      <td>0.02849</td>\n",
       "      <td>0.00204</td>\n",
       "      <td>0.04410</td>\n",
       "      <td>0.01153</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00136</td>\n",
       "      <td>0.01696</td>\n",
       "      <td>0.03596</td>\n",
       "      <td>0.01221</td>\n",
       "      <td>0.04545</td>\n",
       "      <td>0.01560</td>\n",
       "      <td>0.04410</td>\n",
       "      <td>0.00271</td>\n",
       "      <td>0.00068</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4862</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.02180</td>\n",
       "      <td>0.01357</td>\n",
       "      <td>0.01543</td>\n",
       "      <td>0.00782</td>\n",
       "      <td>0.01111</td>\n",
       "      <td>0.01028</td>\n",
       "      <td>0.01193</td>\n",
       "      <td>0.02283</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00596</td>\n",
       "      <td>0.01974</td>\n",
       "      <td>0.02489</td>\n",
       "      <td>0.03126</td>\n",
       "      <td>0.02036</td>\n",
       "      <td>0.02242</td>\n",
       "      <td>0.02468</td>\n",
       "      <td>0.00391</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1915</td>\n",
       "      <td>0.01775</td>\n",
       "      <td>0.02245</td>\n",
       "      <td>0.01619</td>\n",
       "      <td>0.00992</td>\n",
       "      <td>0.01567</td>\n",
       "      <td>0.01358</td>\n",
       "      <td>0.00940</td>\n",
       "      <td>0.01723</td>\n",
       "      <td>0.02402</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00366</td>\n",
       "      <td>0.01410</td>\n",
       "      <td>0.01671</td>\n",
       "      <td>0.03760</td>\n",
       "      <td>0.01932</td>\n",
       "      <td>0.03029</td>\n",
       "      <td>0.03446</td>\n",
       "      <td>0.00261</td>\n",
       "      <td>0.00157</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22831</td>\n",
       "      <td>0.02816</td>\n",
       "      <td>0.01371</td>\n",
       "      <td>0.00767</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0.01380</td>\n",
       "      <td>0.00548</td>\n",
       "      <td>0.00473</td>\n",
       "      <td>0.02076</td>\n",
       "      <td>0.02716</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00604</td>\n",
       "      <td>0.01494</td>\n",
       "      <td>0.01734</td>\n",
       "      <td>0.04148</td>\n",
       "      <td>0.02483</td>\n",
       "      <td>0.03359</td>\n",
       "      <td>0.03679</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00044</td>\n",
       "      <td>0.00131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 65 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Ncodons      UUU      UUC      UUA      UUG      CUU      CUC      CUA  \\\n",
       "0     1995  0.01654  0.01203  0.00050  0.00351  0.01203  0.03208  0.00100   \n",
       "1     1474  0.02714  0.01357  0.00068  0.00678  0.00407  0.02849  0.00204   \n",
       "2     4862  0.01974  0.02180  0.01357  0.01543  0.00782  0.01111  0.01028   \n",
       "3     1915  0.01775  0.02245  0.01619  0.00992  0.01567  0.01358  0.00940   \n",
       "4    22831  0.02816  0.01371  0.00767  0.03679  0.01380  0.00548  0.00473   \n",
       "\n",
       "       CUG      AUU  ...      CGG      AGA      AGG      GAU      GAC  \\\n",
       "0  0.04010  0.00551  ...  0.00451  0.01303  0.03559  0.01003  0.04612   \n",
       "1  0.04410  0.01153  ...  0.00136  0.01696  0.03596  0.01221  0.04545   \n",
       "2  0.01193  0.02283  ...  0.00596  0.01974  0.02489  0.03126  0.02036   \n",
       "3  0.01723  0.02402  ...  0.00366  0.01410  0.01671  0.03760  0.01932   \n",
       "4  0.02076  0.02716  ...  0.00604  0.01494  0.01734  0.04148  0.02483   \n",
       "\n",
       "       GAA      GAG      UAA      UAG      UGA  \n",
       "0  0.01203  0.04361  0.00251  0.00050  0.00000  \n",
       "1  0.01560  0.04410  0.00271  0.00068  0.00000  \n",
       "2  0.02242  0.02468  0.00391  0.00000  0.00144  \n",
       "3  0.03029  0.03446  0.00261  0.00157  0.00000  \n",
       "4  0.03359  0.03679  0.00000  0.00044  0.00131  \n",
       "\n",
       "[5 rows x 65 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7016f3ee",
   "metadata": {},
   "source": [
    "### Splitting dataset into training and testing sets using Stratified strategy (70-30 split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "93ccc1c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "train_features, test_features, train_labels, test_labels = train_test_split(features, \n",
    "                                                                            labels, \n",
    "                                                                            test_size=0.3,\n",
    "                                                                            stratify=labels,\n",
    "                                                                            random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b7f97a6d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "====================\n",
      " Train-Test Split\n",
      "====================\n",
      "\n",
      "Overall class ratio:\n",
      "Kingdom\n",
      "bct        0.336953\n",
      "vrl        0.327019\n",
      "pln        0.175927\n",
      "inv        0.106503\n",
      "vrt        0.053598\n",
      "dtype: float64\n",
      " \n",
      "Train set class ratio:\n",
      "Kingdom\n",
      "bct        0.337019\n",
      "vrl        0.326952\n",
      "pln        0.175937\n",
      "inv        0.106453\n",
      "vrt        0.053639\n",
      "dtype: float64\n",
      " \n",
      "Test set class ratio:\n",
      "Kingdom\n",
      "bct        0.336798\n",
      "vrl        0.327175\n",
      "pln        0.175905\n",
      "inv        0.106620\n",
      "vrt        0.053503\n",
      "dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print('====================')\n",
    "print(' Train-Test Split')\n",
    "print('====================')\n",
    "print('')\n",
    "print('Overall class ratio:')\n",
    "print('{}'.format(labels.value_counts() / len(labels)))\n",
    "print(' ')\n",
    "print('Train set class ratio:')\n",
    "print('{}'.format(train_labels.value_counts() / len(train_labels)))\n",
    "print(' ')\n",
    "print('Test set class ratio:')\n",
    "print('{}'.format(test_labels.value_counts() / len(test_labels)))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mlbd",
   "language": "python",
   "name": "mlbd"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
